data/src/main/java/org/apache/iceberg/data/DeleteFilter.java

```diff
@@ -275,17 +273,12 @@ private CloseableIterable<Record> openDeletes(DeleteFile deleteFile, Schema dele
     InputFile input = getInputFile(deleteFile.path().toString());
     switch (deleteFile.format()) {
       case AVRO:
-        return Avro.read(input)
-            .project(deleteSchema)
-            .reuseContainers()
-            .createReaderFunc(DataReader::create)
-            .build();
+        return Avro.read(input).project(deleteSchema).createReaderFunc(DataReader::create).build();
 
       case PARQUET:
         Parquet.ReadBuilder builder =
             Parquet.read(input)
                 .project(deleteSchema)
-                .reuseContainers()
```

Comment:
Why not disable ORC as well?

Reply:
Yes, but ByteBuffer isn't one of those and it's ByteBuffer that is used for binary fields.
```
  public static class BytesReader extends PrimitiveReader<ByteBuffer> {
    public BytesReader(ColumnDescriptor desc) {
      super(desc);
    }

    @Override
    public ByteBuffer read(ByteBuffer reuse) {
      Binary binary = column.nextBinary();
      ByteBuffer data = binary.toByteBuffer();
      if (reuse != null && reuse.hasArray() && reuse.capacity() >= data.remaining()) {
        data.get(reuse.array(), reuse.arrayOffset(), data.remaining());
        reuse.position(0);
        reuse.limit(binary.length());
        return reuse;
      } else {
        byte[] array = new byte[data.remaining()];
        data.get(array, 0, data.remaining());
        return ByteBuffer.wrap(array);
      }
    }
  }
  ```

There's also the fact that even if the structure were reused here for performance, it'll still be copied a few lines above.

I'd argue that the `Record::copy` is also broken, because I assume it's expected to do a deep copy. I worry the changes needed to ensure it's working correctly would be much more bigger, perhaps intractable, and carry larger risks.

Reply:
This is a great find.

I think it would be a reasonable alternative to fix GenericRecord.copy as well, for ByteBuffers.  (looks like its assuming primitives, which ByteBuffer is not..).

Re:  reuseContainer, point taken that we may probably incur the similar cost if we are to copy.  But I am not familiar with all the cases though, how about perf of nested columns (or would that also break due to GenericRecord copy?)   My observation being, we generally dont put reuseContainers that much in the code.

I would like to see the opinion of @rdblue @RussellSpitzer  as well, as I think Ryan wrote the original GenericRecord.

Reply:
https://github.com/apache/iceberg/issues/6685 I think is related to the copy issue. I don't think we should be disabling container re-use over all but I do think the user of the reader should make copies if required. I'll need to read through the whole pr later though